pkg/kubelet/cm/qos_container_manager_linux.go

```diff
@@ -106,9 +113,90 @@ func (m *qosContainerManagerImpl) Start(nodeInfo *v1.Node, activePods ActivePods
+	// update the QoS reserve on pod entry/exit. This is really only
+	// needed in the case that one of the QoS cgroups is over its
+	// resoruce limit and we are wanting to put continuous downward
+	// pressure on the cgroup such that we can eventually get to the
+	// desired limit.
+	go wait.Until(func() {
+		err := m.UpdateCgroups()
+		if err != nil {
+			glog.Warningf("[ContainerManager] Failed to reserve QoS requests: %v", err)
+		}
+	}, 30*time.Second, wait.NeverStop)
+
 	return nil
 }
 
+// setMemoryReserve sums the memory limits of all pods in a QOS class,
+// calculates QOS class memory limits, and set those limits in the
+// CgroupConfig for each QOS class.
+func (m *qosContainerManagerImpl) setMemoryReserve(configs map[v1.PodQOSClass]*CgroupConfig, percentReserve int64) {
+	qosMemoryRequests := map[v1.PodQOSClass]int64{
+		v1.PodQOSGuaranteed: 0,
+		v1.PodQOSBurstable:  0,
+	}
+
+	// Sum the pod limits for pods in each QOS class
+	pods := m.activePods()
+	for _, pod := range pods {
+		podMemoryRequest := int64(0)
+		qosClass := qos.GetPodQOS(pod)
+		if qosClass == v1.PodQOSBestEffort {
+			// limits are not set for Best Effort pods
+			continue
+		}
+		for _, container := range pod.Spec.Containers {
+			podMemoryRequest += container.Resources.Requests.Memory().Value()
+		}
+		qosMemoryRequests[qosClass] += podMemoryRequest
+	}
+
+	allocatable := m.nodeInfo.Status.Allocatable.Memory().Value()
+	if allocatable == 0 {
+		glog.V(2).Infof("[Container Manager] Memory capacity reported as 0, might be in standalone mode.  Not setting QOS memory limts.")
+		return
+	}
+
+	for qos, limits := range qosMemoryRequests {
+		glog.V(2).Infof("[Container Manager] %s pod requests total %d bytes (reserve %d%%)", qos, limits, percentReserve)
+	}
+
+	// Calculate QOS memory limits
+	burstableLimit := allocatable - (qosMemoryRequests[v1.PodQOSGuaranteed] * percentReserve / 100)
+	bestEffortLimit := burstableLimit - (qosMemoryRequests[v1.PodQOSBurstable] * percentReserve / 100)
+	configs[v1.PodQOSBurstable].ResourceParameters.Memory = &burstableLimit
+	configs[v1.PodQOSBestEffort].ResourceParameters.Memory = &bestEffortLimit
+}
+
+// retrySetMemoryReserve checks for any QoS cgroups over the limit
+// that was attempted to be set in the first Update() and adjusts
+// their memory limit to the usage to prevent further growth.
+func (m *qosContainerManagerImpl) retrySetMemoryReserve(configs map[v1.PodQOSClass]*CgroupConfig, percentReserve int64) {
+	// Unreclaimable memory usage may already exceeded the desired limit
+	// Attempt to set the limit near the current usage to put pressure
+	// on the cgroup and prevent further growth.
+	for qos, config := range configs {
+		stats, err := m.cgroupManager.GetResourceStats(config.Name)
+		if err != nil {
+			glog.V(2).Infof("[Container Manager] %v", err)
+			return
+		}
+		usage := stats.MemoryStats.Usage
+
+		// Because there is no good way to determine of the original Update()
```

Comment:
Curious: Why not read the value in `memory.limit_in_bytes` to validate it against the expectations?